lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/DelegationSpec.hs

```diff
@@ -0,0 +1,423 @@
+
+instance Buildable Env where
+    build env = "Env:\n" <> blockMapF
+        [ ("wallet" :: String, build $ wallet env)
+        , ("txs", blockListF' "-" txF $ reverse $ txs env)
+        , ("rejected txs", blockListF' "-" build $ reverse $ rejectedTxs env)
+        , ("ledger", build $ ledger env)
+        ]
+      where
+        txF tx = build (txid tx) <> ": " <> build tx
+
+
+rewardAccountF :: RewardAccount -> Builder
+rewardAccountF (RewardAccount a) = build (B8.unpack a)
+
+
+instance Buildable Cert where
+    build (RegisterKey k) = "RegKey " <> rewardAccountF k
+    build (Delegate k) = "Deleg " <> rewardAccountF k
+    build (DeRegisterKey k) = "DeReg " <> rewardAccountF k
+
+instance Arbitrary Cmd where
+    -- We don't want to generate too many adversarial registrations (we don't
+    -- expect them to happen in reality), but at least some, and enough to cause
+    -- consistent failures if something is wrong.
+    arbitrary = frequency
+        [ (85, CmdSetPortfolioOf . getNonNegative <$> arbitrary)
+        , (5, CmdAdversarialReg <$> arbitrary)
+        , (10, pure CmdOldWalletToggleFirstKey)
+        ]
+    shrink = genericShrink
+
+data Env = Env
+    { wallet :: DelegationState StakeKey'
+    , ledger :: Ledger
+    , txs :: [Tx]
+        -- ^ All accepted transactions so far in the chain. Newest first.
+    , rejectedTxs :: [String]
+        -- ^ User-generated txs that were rejected. I.e. not adversarial ones.
+        -- Newest first.
+    } deriving (Show, Eq)
+
+env0 :: Env
+env0 = Env (initialDelegationState accK) initialLedger [] []
+
+stepCmd :: Cmd -> Env -> Env
+stepCmd (CmdSetPortfolioOf n) env =
+    case setPortfolioOf (wallet env) mkAddr (acctIsReg (ledger env)) n of
+        Just tx -> tryApplyTx tx env
+        Nothing -> env
+  where
+    mkAddr k = Address $ "addr" <> unRewardAccount (toRewardAccount k)
+
+stepCmd (CmdAdversarialReg k) env
+    | k `Set.member` regs (ledger env)
+        = env -- We don't want tx errors to appear in @rejectedTxs@.
+    | otherwise
+        = tryApplyTx (Tx [RegisterKey k] [] []) env
+stepCmd CmdOldWalletToggleFirstKey env =
+            let
+                key0 = toRewardAccount (keyAtIx (wallet env) minBound)
+                isReg =  key0 `Set.member` (regs (ledger env))
+                tx = Tx [if isReg then DeRegisterKey key0 else RegisterKey key0] [] []
+            in tryApplyTx tx env
+
+tryApplyTx :: Tx -> Env -> Env
+tryApplyTx tx env = case ledgerApplyTx tx h (ledger env) of
+    Right l' -> env
+        { ledger = l'
+        , wallet = applyTx tx (txid tx) $ wallet env
+        , txs = tx : txs env
+        }
+    Left e -> env { rejectedTxs = e : rejectedTxs env }
+  where
+    h = txid tx
+
+
+applyCmds :: Env -> [Cmd] -> Env
+applyCmds = foldl (flip stepCmd)
+
+applyTxs :: Env -> [Tx] -> Env
+applyTxs = foldl (flip tryApplyTx)
+
+txid :: Tx -> Hash "Tx"
+txid = Hash . BS.take 4 . blake2b224 . B8.pack . show
```

Comment:
At one point I wanted a lookup-map to assign txs with sequentially incrementing ids — like tx0, tx1, tx2 — but didn't end up doing it. As long as they are formatted nicely, short hashes are fairly readable.